smart-contracts/wasm-chain-integration/src/v1/mod.rs

```diff
@@ -486,6 +522,43 @@ mod host {
                 );
                 Ok(Interrupt::QueryExchangeRates.into())
             }
+            CHECK_ACCOUNT_SIGNATURE_TAG if support_account_signature_checks => {
+                ensure!(
+                    length >= ACCOUNT_ADDRESS_SIZE,
+                    "Account balance queries must have exactly 32 bytes of payload, but was {}",
+                    length
+                );
+                // Overflow is not possible in the next line on 64-bit machines.
+                ensure!(start + length <= memory.len(), "Illegal memory access.");
+                let mut addr_bytes = [0u8; ACCOUNT_ADDRESS_SIZE];
+                addr_bytes.copy_from_slice(&memory[start..start + ACCOUNT_ADDRESS_SIZE]);
+                let address = AccountAddress(addr_bytes);
+                if energy.tick_energy(constants::copy_parameter_cost(length_u32)).is_err() {
+                    bail!(OutOfEnergy);
+                }
```

Comment:
We could check this before copying the account address so we avoid doing unnecessary work when the energy is too low.

Reply:
I see that we don't charge for moving the account keys below, so perhaps this is consistent afterall.

Reply:
That's a good point, we should do it before when we can.

Reply:
That's changed now.